### PR TITLE
suggesting a pull-request-template.

### PR DESCRIPTION
**Pull request by mattulbrich**
## Intended Change
There is no guided process for writing PRs. But there should be. Therefore: Suggesting this template here.

## Reason for the pull request

improving git/github maintenance workflow.

## Touched elements

adding file to `.github`.
     
## Information for the reviewer
    
I tested that a file in this place should in the future appear as pull-request template on an empty repo of my own. I assume it will work here to.

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+## Intended Change
+<!-- Please give a brief description of what behaviour changes and 
+     why it should be changed. -->
+
+## Reason for the pull request
+
+<!-- one item like: 
+     bugfix, new feature, refactoring, repaired workflow, ... -->
+
+## Touched elements
```
Comment (wadoon) [inline .github/pull_request_template.md:23]: I see this as redundant information. The same is tracked by Github under "Files changed".

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+<!-- A brief overview of the touched code areas. No need to list 
+     files (that is available in github). Mention things like:
+     TacletApps in key.core, ANTLR for JML, github config, ... -->
+     
+## Information for the reviewer
```
Comment (wadoon) [inline .github/pull_request_template.md:29]: Unclear what to tell to the reviewers. 

What would make us a review easier?

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+## Additional contact(s)
+
+<!-- who apart from yourself is involved in this pull request?
+     Use @mentions to refer to them here -->
+     
+<!-- DRAFT MODE: Please note that on the button to submit this pull
+     request you can select between submitting a merge-ready request
+     or one in draft mode (still evolving). 
+     Please use the draft mode unless you think that your proposal
+     should be brought onto master in the current form. -->
```
Comment (wadoon) [inline .github/pull_request_template.md:42]: Maybe we want additional information, e.g., to create a good changelog.

https://github.com/stevemao/github-issue-templates/blob/master/checklist2/PULL_REQUEST_TEMPLATE.md

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+
+<!-- Thanks for submitting this pull request for KeY -->
```
Comment (wadoon) [inline .github/pull_request_template.md:2]: I would also add a note:

> By making this pull request, your contribution will be licensed under GPL2-or-later.

File: .github/pull_request_template.md (hunk 3 above)
Comment (wadoon) [reply .github/pull_request_template.md:42]: Also required may be the consideration: "The documentation is/will be updated".

File: key.util/src/main/java/org/key_project/util/helper/FindResources.java [hunk 5]
```diff
@@ -56,7 +56,7 @@ public static <T> List<Path> getResources(String path, Class<T> clazz)
             Path dir = fs.getPath(path);
             return Files.list(dir).collect(Collectors.toList());
         }
-        throw new UnsupportedOperationException("Cannot list files for URL $dirURL");
```
Comment (wadoon) [inline key.util/src/main/java/org/key_project/util/helper/FindResources.java:59]: Please revert these unrelated changes.

**Review by wadoon — CHANGES_REQUESTED** [review]

File: .github/pull_request_template.md (hunk 1 above)
Comment (mattulbrich) [reply .github/pull_request_template.md:23]: Not quite as mentioned then. But we can drop that.

File: key.util/src/main/java/org/key_project/util/helper/FindResources.java (hunk 5 above)
Comment (mattulbrich) [reply key.util/src/main/java/org/key_project/util/helper/FindResources.java:59]: Sure, how did they get in in the first place?!

File: .github/pull_request_template.md (hunk 4 above)
Comment (mattulbrich) [reply .github/pull_request_template.md:2]: Is this necessary? Contributing to a GPL project seems to imply this automatically, I think.

File: .github/pull_request_template.md (hunk 3 above)
Comment (mattulbrich) [reply .github/pull_request_template.md:42]: I do not really believe in Changelogs. Is that not usually a very obvious reason for merge conflicts?

What does "The documentation is/will be updated" mean? I do not understand this point.

File: .github/pull_request_template.md (hunk 4 above)
Comment (wadoon) [reply .github/pull_request_template.md:1]: I am not sure. But the a little difference is, that now the author also acknowledges a re-licensing to GPL3.

Comment (mi-ki) [reply .github/pull_request_template.md:1]: It's in the [GitHub Terms of Service](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#6-contributions-under-repository-license). I am pretty sure that every user on GitHub needed to accept those.

Comment (mi-ki) [reply .github/pull_request_template.md:1]: If anything, we can add a link to our license. However, a hard-coded (static) reference is not a good idea, in my opinion.

Comment (mattulbrich) [reply .github/pull_request_template.md:1]: My vote would be to omit this here if possible.

**Comment by mattulbrich** [comment]
I have just noticed that the text of the PR will be used for the merge commit message. Perhaps it is not such a good idea to have all these xml comments after all. I suggest one modifies them to `>` markings like in the bug issue template to avoid cluttering the commit messages. Right?

File: .github/pull_request_template.md (hunk 3 above)
Comment (wadoon) [reply .github/pull_request_template.md:50]: > What does "The documentation is/will be updated" mean? I do not understand this point.
The author adds a note, whether this Pull Request results are also reflected in `key-docs`. 

I think changelog are important for a real project (just look at dafny, z3, cbmc, ...).

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] Refactoring (behaviour should not change or only minimally change)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Breaking change (fix or feature that would cause existing functionality to change)
+- [ ] There are changes to the (Java) code
+- [ ] There are changes to the taclet rule base
+- [ ] There are changes to the deployment/CI infrastructure (gradle, github, ...)
+- [ ] Other: 
+
+## Ensuring quality
```
Comment (wadoon) [inline .github/pull_request_template.md:31]: This section is kind of strange. 

Some of these things are hard to check ("runtime performance"). On some other we can discuss for weeks ("javadoc"). And other things like "I wrote a something in key-docs" are not reflected. 

I would such get rid of these points. They do not bring anyone further, or makes the review easier. 

I think this templating becomes to "German" and remembers me on the stupid ISO27001-driven forms back in my apprenticeship. 

Things that make this review easier is something like: "I needed to refactor the method class#method by adding a new parameter. Following classes are automatically updated w/o further intentions."

Comment (mattulbrich) [reply .github/pull_request_template.md:31]: I took the idea of a `[ ]` list from a reference you suggested earlier, @wadoon.
I am happy with it and without it. Could you edit it?

**Comment by mattulbrich** [comment]
> Personally, I do mostly not see a problem in commit messages that are too long (usually, rather the opposite), but I can see why others might feel different. Yet, upon a quick scanning, I did not find any commit messages that were too cluttered in my eyes. Can you maybe point to any specific examples which you deem too cluttered?

Long messages are fine, but if the have all the `<!-- comments -->` inside them, it can be annoying.

As you can see for 4194ee36c254cca96bc3ecd294f171a8ec1eed9a, the entire PR description (by default) becomes part of the commit message (not the first line ...) If that is full of comments: no good.

**Comment by mi-ki** [comment]
> > Personally, I do mostly not see a problem in commit messages that are too long (usually, rather the opposite), but I can see why others might feel different. Yet, upon a quick scanning, I did not find any commit messages that were too cluttered in my eyes. Can you maybe point to any specific examples which you deem too cluttered?
> 
> Long messages are fine, but if the have all the `<!-- comments -->` inside them, it can be annoying.
> 
> As you can see for [4194ee3](https://github.com/KeYProject/key/commit/4194ee36c254cca96bc3ecd294f171a8ec1eed9a), the entire PR description (by default) becomes part of the commit message (not the first line ...) If that is full of comments: no good.

Thanks for the pin-point. So your point is that the comments should not be included after the PR description. Hmm, I somehow do not find that particularly problematic, but rather good for transparency, e.g., to understand the requested changes or maybe any particularities about the specific PR. Moreover, the main information is still to be found directly at the top, such as the title and the description in the first comment, and most people would probably not scroll down to the comments below.

Yet, my opinion is not too strong on that point. Do you see cause for action?

**Review by mi-ki — APPROVED** [review]

**Review by mi-ki — APPROVED** [review]

**Review by mi-ki — APPROVED** [review]

**Comment by mattulbrich** [comment]
Let me refactor w/o comments.

**Comment by FliegendeWurst** [comment]
It is possible to configure the merge commit message:

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+     Who apart from yourself is involved in this pull request?
+     Use @mentions to refer to them here -->
+     
+<!-- DRAFT MODE: Please note that on the button to submit this pull
+     request you can select between submitting a merge-ready request
+     or one in draft mode (still evolving). 
+     Please use the draft mode unless you think that your proposal
+     should be brought onto master in the current form. -->
+
+The contributions within this pull request will be licensed under GPL2-or-later wihtin KeY.
```
Comment (FliegendeWurst) [inline .github/pull_request_template.md:52]:
```suggestion
The contributions within this pull request will be licensed under GPLv2 (only) wihtin KeY.
```
See https://github.com/KeYProject/key/discussions/3212#discussioncomment-6538678

File: .github/pull_request_template.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+- [ ] There are changes to the taclet rule base
+- [ ] There are changes to the deployment/CI infrastructure (gradle, github, ...)
+- [ ] Other: 
+
+## Ensuring quality
+    
+- [ ] I made sure that introduced/changed code has been well documented (javadoc).
+- [ ] I added new test case(s) for new functionality.
+- [ ] I have tested the feature as follows: ...
+- [ ] I have checked that runtime performance has not deteriorated
```
Comment (FliegendeWurst) [inline .github/pull_request_template.md:36]:
```suggestion
- [ ] I made sure that introduced/changed code is well documented (javadoc and inline comments).
- [ ] I made sure that new/changed end-user features are well documented (https://github.com/KeYProject/key-docs).
- [ ] I added new test case(s) for new functionality.
- [ ] I have tested the feature as follows: ...
- [ ] I have checked that runtime performance has not deteriorated.
```

File: .github/pull_request_template.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+- [ ] I have tested the feature as follows: ...
+- [ ] I have checked that runtime performance has not deteriorated  
+
+## Additional information and contact(s)
+
+<!-- Add further information to help the reviewer understand the request.
+     Leave empty if you are sure the reviewer does not need more
+     
+     Who apart from yourself is involved in this pull request?
+     Use @mentions to refer to them here -->
```
Comment (FliegendeWurst) [inline .github/pull_request_template.md:44]:
```suggestion
     Who apart from yourself is involved in this pull request?
     Use @mentions to refer to them here.
     Use Co-Authored-By in your git commits if applicable. -->
```

File: .github/pull_request_template.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+<!-- Since the project has a strict review policy, please make the -->
+<!-- reviewer's job easier by providing the necessary information -->
+<!-- in the text below. The comments may remain since they will be -->
+<!-- invisible when showing the PR. -->
+
+## Related Issue
+
+<!-- Please remove if this PR is not related to an issue. -->
+<!-- Please add number if it is in answer to an issue. -->
+This pull request answers to issue #.
```
Comment (FliegendeWurst) [inline .github/pull_request_template.md:11]:
```suggestion
This pull request addresses #.
```

**Review by WolframPfeifer — APPROVED** [review]
I am quite happy with this template in its current state. In particular, I like the list `Ensuring quality`. However, I would not expect that every PR has all the boxes checked, but rather understand it as a reminder for the author(s) of the PR what they should consider.

Are we missing anything in this template, or can it be merged?

File: .github/pull_request_template.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
+<!-- Thanks for submitting this pull request for KeY -->
+<!-- Since the project has a strict review policy, please make the -->
+<!-- reviewer's job easier by providing the necessary information -->
+<!-- in the text below. The comments may remain since they will be -->
+<!-- invisible when showing the PR. -->
```
Comment (FliegendeWurst) [inline .github/pull_request_template.md:5]:
```suggestion
<!--
Thanks for submitting this pull request for KeY.
Since the project has a strict review policy, please make the
reviewer's job easier by providing the necessary information
in the text below. The comments may remain since they will be
invisible when showing the PR.
-->
```
Easier to read (same content). Should ideally be consistent across the template.

**Review by FliegendeWurst — COMMENTED** [review]
Just a formatting suggestion.

**Review by wadoon — APPROVED** [review]
